src/plugins/explore/public/components/tabs/tabs.tsx

```diff
@@ -21,12 +23,17 @@ import { ExploreServices } from '../../types';
  */
 export const ExploreTabsComponent = () => {
   const dispatch = useDispatch();
-  // Get services from context
   const { services } = useOpenSearchDashboards<ExploreServices>();
   const registryTabs = services.tabRegistry.getAllTabs();
-
+  const results = useSelector((state: RootState) => state.results);
   const activeTabId = useSelector(selectActiveTab);
 
+  useEffect(() => {
+    if (activeTabId === '' && results && Object.keys(results).length > 0) {
+      dispatch(detectAndSetOptimalTab({ services }));
+    }
+  }, [activeTabId, results, dispatch, services]);
```

Comment:
is it possible to declaratively dispatch optimal tab within the executeQueries? I want to avoid `useEffect` unless absolutely necessary, and the logic here looks like it can be done without this sideeffect

Reply:
+1